packages/core/src/modules/oob/OutOfBandModule.ts

```diff
@@ -0,0 +1,181 @@
+    // Either way, we need to get routing
+    const mediationRecord = await this.mediationRecipientService.discoverMediation()
+    const routing = await this.mediationRecipientService.getRouting(mediationRecord)
+    const { connectionRecord: connectionRecord } = await this.connectionService.createInvitation({
+      routing,
+    })
+
+    const outOfBandMessage = new OutOfBandMessage(options)
+    outOfBandMessage.accept.push('didcomm/aip2;env=rfc587')
+    outOfBandMessage.accept.push('didcomm/aip2;env=rfc19')
+    supportedHandshakeProtocols.forEach((p) => outOfBandMessage.addHandshakeProtocol(p))
+    connectionRecord.didDoc.didCommServices.forEach((s) => outOfBandMessage.addService(s))
+
+    return { outOfBandMessage, connectionRecord }
+  }
+
+  /**
+   * Creates a connection record based on out-of-band message.
+   *
+   * @param outOfBandMessage
+   * @param config
+   * @returns Connection record
+   */
+  public async receiveInvitation(
+    outOfBandMessage: OutOfBandMessage,
+    config: { autoAccept: boolean }
+  ): Promise<ConnectionRecord> {
+    if (!outOfBandMessage.handshakeProtocols || outOfBandMessage.handshakeProtocols.length === 0) {
+      throw new AriesFrameworkError('Missing required `handshake_protocols` attribute in OOB message.')
+    }
+
+    const mediationRecord = await this.mediationRecipientService.discoverMediation()
+    const routing = await this.mediationRecipientService.getRouting(mediationRecord)
+    const invitation = new ConnectionInvitationMessage({ label: 'connection label', ...outOfBandMessage.services[0] })
+
+    let connectionRecord = await this.connectionService.processInvitation(invitation, { routing })
+    if (config.autoAccept) {
+      connectionRecord = await this.acceptInvitation(connectionRecord.id)
+    }
+
+    if (outOfBandMessage.getRequests()) {
+      throw new AriesFrameworkError('OOB invitation contains unsupported `request~attach` attribute.')
+    }
+
+    return connectionRecord
+  }
+
+  /**
+   * Creates an out-of-band message and adds given agent message to `requests~attach` attribute.
+   *
+   * @param message A message that will be send inside out-of-band message
+   * @param options Optinal attributes contained in out-of-band message
+   * @returns Out-of-band message
+   */
+  public async createOobMessage(message: AgentMessage, options: OutOfBandMessageOptions) {
+    if (!message.service) {
+      throw new AriesFrameworkError(
+        `Out of band message with id ${message.id} and type ${message.type} does not have a ~service decorator`
+      )
+    }
+
+    const outOfBandMessage = new OutOfBandMessage(options)
+    outOfBandMessage.accept.push('didcomm/aip2;env=rfc587')
+    outOfBandMessage.accept.push('didcomm/aip2;env=rfc19')
+
+    // To support newer OOB messages we need to add service from message and then remove `~service` attribute from message
+    outOfBandMessage.services.push(message.service.toDidCommService())
+    message.service = undefined
```

Comment:
I think the advantage of using OOB is that the `AgentMessage` itself doesn't need the `~service` anymore.

Reply:
This is just a workaround for now. It allows us to pass the `AgentMessage` into `receiveMessage` and still be able to process it with the current implementation.